loadgen/loadgen.cc

```diff
@@ -1246,6 +1292,29 @@ void QuerySamplesComplete(QuerySampleResponse* responses, size_t response_count,
     query->response_delegate->SampleComplete(sample, response, timestamp,
                                              response_cb);
   }
+  // PerfClock::time_point end_timestamp = PerfClock::now();
+  // mlperf::samples_overhead_acum += (end_timestamp - timestamp).count();
+}
+
+void FirstTokenComplete(QuerySampleResponse* responses, size_t response_count,
```

Comment:
Hi @pgmpablo157321 qq: In accuracy mode in the Server scenario, is the first token output being logged? I checked and haven't seen it so far, looking at `TokenComplete`.

Or is the client expected to resend the first token together with the remainder tokens when they call `QuerySamplesComplete` for the query?
If not, then somehow we need to find a way to record it in `TokenComplete` or perhaps find a way to store it and use later [when loging accuracy data](https://github.com/mlcommons/inference/blob/51b2203fea57fe9e06f95902497ea31d10d6159e/loadgen/loadgen.cc#L121-L125) after completion

Reply:
I think we should record it in the `TokenComplete` function when it is running accuracy mode. It seems it is only a couple of lines so I can add them. Do you think we should log this in the same accuracy file or should it be another one?

Reply:
We currently use mlperf_log_accuracy.json and it's list of {"seq_id", "qsl_idx", "data"}. I wonder adding "token_data" into this would work out nicely like {"seq_id", "qsl_idx", "data", "token_data"}, if and only if we use token latency measurement.

Reply:
Sounds reasonable, I think we can cache this value when the `TokenComplete` function is called

Reply:
@pgmpablo157321 @nv-jinhosuh The extra field for the "first token" is actually a good idea. 
Also, from taskforce meeting today, it appears when client calls `FirstTokenComplete` they'll have to send the first two tokens instead of just the first token, because we discovered that the first generated token is always the same (`29871`) for any input, so adding the second token ensures a client doesn't spuriously presend `29871` to beat the TTFT constraint.
I believe everything else in the loadgen should remain same, unless `TokenComplete` has hardcoded an expectation of a single token in the response data...

Reply:
Oh woops. I think I heard it from @nvzhihanj . We probably have to call token complete twice then, and store `29871`+`real first token`, and mark the token size to be as such. Thanks @attafosu for letting me know.

Reply:
@attafosu agreed. @nvzhihanj  does it sound better? Or we can just send second token as first token is always `29871` and the validity (of `29871` and `second token`) can be checked by comparing this against the full query sample response when the LoadGen generates the accuracy dump JSON as I explained above besides the `token_data` extra section: https://github.com/mlcommons/inference/pull/1523#discussion_r1415750650

Reply:
> I believe everything else in the loadgen should remain same, unless TokenComplete has hardcoded an expectation of a single token in the response data...

@attafosu @nv-jinhosuh @nvzhihanj 
It was not hardcoded, I suggest to send the first two tokens and call `FirstTokenComplete` just once

Reply:
@pgmpablo157321 I think I like this as it's not making the issue overcomplicated and captures an aspect of first token latency + one token generation latency. One can still interpolate real first token latency by using this together with TPOT I guess. I would still think some submitters may not like this idea so perhaps this is one to be discussed in the WG.

Reply:
> Hi @pgmpablo157321 qq: In accuracy mode in the Server scenario, is the first token output being logged? I checked and haven't seen it so far, looking at TokenComplete.

Now it is being logged!